### ISSUE TITLE
Sometimes failing to typecheck template parameters inside record types

### ISSUE THREAD

**Issue by lauraharker** [google/closure-compiler#2867]
The following code does not cause a warning even though `Foo.prototype.foo`'s return type appears to be incorrect.
``` javascript
/**
 * @interface
 * @template T
 */
function Int() {}

/** @return {{prop: T}} */
Int.prototype.foo = function() {};

/**
 * @constructor
 * @implements {Int<string>}
 */
function Foo() {}

/**
 * @return {{prop: number}} // Should be {prop: string}, but no warning
 * @override
 */
Foo.prototype.foo = function() {}
```

https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%250A%2520*%2520%2540interface%250A%2520*%2520%2540template%2520T%250A%2520*%252F%250Afunction%2520Int()%2520%257B%257D%250A%250A%252F**%2520%2540return%2520%257B%257Bprop%253A%2520T%257D%257D%2520*%252F%250AInt.prototype.foo%2520%253D%2520function()%2520%257B%257D%253B%250A%250A%252F**%250A%2520*%2520%2540constructor%250A%2520*%2520%2540implements%2520%257BInt%253Cstring%253E%257D%250A%2520*%252F%250Afunction%2520Foo()%2520%257B%257D%250A%250A%252F**%250A%2520*%2520%2540return%2520%257B%257Bprop%253A%2520number%257D%257D%2520%252F%252F%2520Should%2520be%2520%257Bprop%253A%2520string%257D%2520%253F%253F%250A%2520*%2520%2540override%250A%2520*%252F%250AFoo.prototype.foo%2520%253D%2520function()%2520%257B%257D%26input1%26conformanceConfig%26externs%26refasterjs-template%26CHECK_SYMBOLS%3Dtrue%26MISSING_PROPERTIES%3Dtrue%26TRANSPILE%3Dtrue%26CHECK_TYPES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue

**Comment by lauraharker** [comment]
Turns out this is because the `TemplateTypeMapReplacer` class extends `ModificationVisitor` and intentionally (?) sets a boolean property in `ModificationVisitor`'s constructor telling it not to visit properties.
https://github.com/google/closure-compiler/blob/4d7a626b9a9796fbce17577c23d92ef2e80a2c5b/src/com/google/javascript/rhino/jstype/TemplateTypeMapReplacer.java#L61

https://github.com/google/closure-compiler/blob/4d7a626b9a9796fbce17577c23d92ef2e80a2c5b/src/com/google/javascript/rhino/jstype/ModificationVisitor.java#L61
